temporian/core/data/dtype.py

```diff
@@ -38,3 +38,51 @@ def MissingValue(dtype: DType) -> Any:
         return ""
     else:
         raise ValueError(f"Non implemented type {dtype}")
+
+
+def get_resulting_dtype(dtype1: DType, dtype2: DType) -> DType:
+    """
+    Get the resulting dtype when combining two arrays with the specified dtypes.
+
+    This function takes two dtypes as input and returns the expected
+    output dtype based on a simplified dtype hierarchy.
+
+    Args:
+        dtype1: The dtype of the first array.
+        dtype2: The dtype of the second array.
+
+    Returns:
+        DType: The resulting dtype
+
+    Raises:
+        ValueError: If either of the input dtypes is not supported.
+    """
+    dtype_hierarchy = {
+        INT32: 1,
+        FLOAT32: 2,
+        INT64: 3,
+        FLOAT64: 4,
+    }
+
+    if dtype1 not in dtype_hierarchy or dtype2 not in dtype_hierarchy:
+        supported_dtypes = ", ".join(dtype_hierarchy.keys())
+        raise ValueError(
+            f"Invalid dtype(s): [{dtype1}, {dtype2}]. Supported dtypes:"
+            f" {supported_dtypes}"
+        )
+
+    # Find the highest hierarchy value between the two input dtypes
+    max_hierarchy = max(dtype_hierarchy[dtype1], dtype_hierarchy[dtype2])
+
+    # handle special case with int64 and float32
+    if (dtype1 == INT64 and dtype2 == FLOAT32) or (
+        dtype1 == FLOAT32 and dtype2 == INT64
+    ):
+        return FLOAT64
```

Comment:
I think this check should go before line 75 (`max_hierarchy = max(dtype_hierarchy[dtype1], ...`)